src/credentials/DeviceAttestationVendorReserved.h

```diff
@@ -0,0 +1,218 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/support/Span.h>
+
+#include <cstdint>
+
+namespace chip {
+namespace Credentials {
+
+struct VendorReservedElement
+{
+    uint16_t vendorId;
+    uint16_t profileNum;
+    uint32_t tagNum;
+    ByteSpan vendorReservedData;
+};
+
+// extract elements out of the device attestation bytespan
+class DeviceAttestationVendorReservedDeconstructor
+{
+
+public:
+    DeviceAttestationVendorReservedDeconstructor() {}
+
+    // read TLV until first profile tag
+    CHIP_ERROR PrepareToReadVendorReservedElements(const ByteSpan & attestationElements, size_t count)
+    {
+        mNumVendorReservedData = count;
+        mAttestationData       = attestationElements;
+
+        tlvReader.Init(mAttestationData);
+        ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+        ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+        // position to first ProfileTag
+        while (true)
+        {
+            ReturnErrorOnFailure(tlvReader.Next());
+            if (!TLV::IsProfileTag(tlvReader.GetTag()))
+                break;
+        }
+        // positioned to first context tag (vendor reserved data)
+        mIsInitialized = true;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Next() { return tlvReader.Next(); }
+
+    size_t GetNumberOfElements() { return mNumVendorReservedData; }
+
+    /**
+     *  @brief Return next VendorReserved element. PrepareToReadVendorReservedElements must be called first.
+     *
+     *  @param[out] element  Next vendor Reserved element
+     *
+     *  @returns   CHIP_NO_ERROR on success
+     *             CHIP_ERROR_INCORRECT_STATE if PrepareToReadVendorReservedElements hasn't been called first
+     *             CHIP_END_OF_TLV if not further entries are present
+     */
+    CHIP_ERROR GetNextVendorReservedElement(struct VendorReservedElement & element)
+    {
+        VerifyOrReturnError(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+
+        while (1)
+        {
+            ReturnErrorOnFailure(Next());
```

Comment:
End of TLV is not a failure. Suggest being clearer about the condition checking on Next(). While you return in all those conditions, some status are not real "failures" and so it makes this code harder to parse.